File: Numbers/number-of-1-bits/ts/hammingWeight.ts
export {}

const testCases: number[] = [11, 128, 2147483645, 0, 1, 2147483647];

//Using bitwise operations
function hammingWeight(n: number): number {
    let count = 0;
    let currentNumber = n;
    while(currentNumber > 0){
        count += currentNumber & 1;
        currentNumber >>>= 1;
    }
    return count
}

for (const n of testCases) {
    console.log(`Input:       ${n} (binary: ${n.toString(2)})`);
    console.log(`Output:      ${hammingWeight(n)}`);
    console.log("-".repeat(35));
}

File: Numbers/number-of-1-bits/go/hamming_weight.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

// Using the Brian Kernighan’s Algorithm I just discovered
func hammingWeight(n int) int {
	var count int = 0
	local := n
	for local != 0 {
		local = local & (local - 1)
		count += 1
	}
	return count
}

func main() {
	testCases := []int{11, 128, 2147483645, 0, 1, 2147483647}

	for _, n := range testCases {
		fmt.Printf("Input:       %d (binary: %s)\n", n, strconv.FormatInt(int64(n), 2))
		fmt.Printf("Output:      %d\n", hammingWeight(n))
		fmt.Println(strings.Repeat("-", 35))
	}
}
